### DISCUSSION TITLE
Where is en_core_web_sm of Python2 for spacy in Python3?

### DISCUSSION THREAD

**Discussion by tengerye** [explosion/spaCy#6733] [Help: Coding & Implementations] [answered]
I need to repeat an experiment. The experiment was conducted in python 2.7 and spacy 1.8.2. The snippet I need to repeat is:
```
import en_core_web_sm

nlp = en_core_web_sm.load()
len(list(nlp.vocab.strings))
```
The output from above is 776980.

Now I would like to repeat the experiment in python 3.7 and spacy 2.3. Where can I find the same size vocabulary?

**Comment by svlandeg** [comment]
I think this explanation from Adriane would be useful to you: https://github.com/explosion/spaCy/issues/5633#issuecomment-648011028

Does that help?

**Comment by tengerye** [comment]
@svlandeg Hi, thank you for your kind reply. I tried the instructions from [#5633 (comment)](https://github.com/explosion/spaCy/issues/5633#issuecomment-648011028).

`len(nlp.vocab.vectors)` gives me 0 and `len(nlp.vocab.strings)` gives me 1160. No it does not work.

**Comment by svlandeg** [comment]
The `_sm` models typically don't have vectors (as opposed to the `_md` and `_lg` models), so it's correct that the count of vectors is 0 in your case.

With respect to the vocabulary size - as explained in that other thread, the vocab isn't prepopulated on loading anymore since spaCy v. 2.3.0. This shouldn't have much impact on your processing though, as lexemes will be typically added to the vocab as you process more texts. You could regard the vocab as just an internal cache.

So I think my main question is: why exactly do you need to replicate the exact number of vocab strings? Are you seeing significant differences when running `nlp(doc)` with the newer `en_core_web_sm`?

**Comment by tengerye** [comment]
@svlandeg Thank you for your kind reply.

The following [snippets](https://github.com/kohpangwei/influence-release/blob/0f656964867da6ddcca16c14b3e4f0eef38a7472/influence/nlprocessor.py#L28) give different outputs:
```
for raw_doc in spam + ham:
    doc = self.nlp(raw_doc)
    docs.append(' '.join(
        [token.lemma_ for token in doc if (token.is_alpha and not (token.is_oov or token.is_stop))]))
```

In the Python 3.7 environment, the `token.is_oov` is always `True`, which leads to empty selection. While spacy 1.8.2 gives reasonable result. The vocabulary is important here for identical repeat.

**Comment by tengerye** [comment]
@svlandeg Hi, sorry to interrupt. But do you have any idea for the approach please?

**Comment by svlandeg** [comment]
Can I ask you to please not ping me just to get your problem higher up in priority? This makes our mentions almost impossible to use. Please note that we are continuously maintaining this issue tracker and the discussion forum, and this is not the only open issue, nor is community support our only job.

To come back to your question:

From the [documentation](https://spacy.io/api/token#attributes): `token.is_oov` refers to whether or not a token has a word vector. This will not be the case for the `en_core_web_sm` model, as I explained before, unless you'd [add vectors](https://spacy.io/usage/vectors-similarity#custom-vectors-add) yourself.

It could very well be that `token.is_oov` used to behave differently. We typically try to avoid breaking changes, but the major release of 2.0 might have revisited some details of these implementations (I wasn't around at the time). You could check through the release notes for this information. Considering Adriane's comment that some of this changed in `2.3`, you might also try a `2.2.x` version of spaCy. 

It doesn't look like this is a bug in spaCy, so I'll move this thread to the discussion forum where others in the community can chime in as well.

**Comment by adrianeboyd** [comment, marked as answer]
In v2.2 and earlier `token.is_oov` referred to whether lexeme data (mainly Brown cluster + prob values) was provided by the model, which is the case for `md` and `lg` models for spacy v2.0-v2.2. In v2.3+, `token.is_oov` refers to whether there's a vector, which are also only in `md` and `lg` models, and the lexeme data is no longer included by default because the features aren't used by the models anymore.

As far as I know, in spacy v1 all models included the lexeme data because the model features were very different.
